File: basic_java_socket/src/main/java/org/example/ChatClient.java
package org.example;
import java.io.*;
import java.net.*;
import java.util.Scanner;

public class ChatClient {
    public static void main(String[] args) {
        try(Socket socket = new Socket("localhost", 12345)) {

            System.out.println("채팅 서버에 접속되었습니다.");

            BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream()));

            new Thread(() -> {
                try {
                    String message;
                    while ((message = reader.readLine()) != null) {
                        System.out.println("수신: " + message);
                    }
                } catch (IOException e) {
                    System.out.println("Connection has been closed.");
                }
            }).start();

            // 사용자 입력을 받아 서버에 메시지 전송
            Scanner scanner = new Scanner(System.in);
            String message;
            while (true) {
                message = scanner.nextLine();
                writer.write(message + "\n");
                writer.flush();
            }
        } catch (IOException e) {
            System.out.println("Connection has been closed.");
        }

    }
}

File: basic_kotlin_socket/src/main/kotlin/ChatClient.kt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.Socket

fun main() = runBlocking {
    val socket = Socket("localhost", 1234)
    val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
    val writer = PrintWriter(socket.getOutputStream(), true)

    GlobalScope.launch(Dispatchers.IO) {
        while (true) {
            val message = reader.readLine()
            if (message != null) {
                println("서버로부터 받은 메시지: $message")
            }
        }
    }

    val consoleReader = BufferedReader(InputStreamReader(System.`in`))
    while (true) {
        val input = consoleReader.readLine()
        writer.println(input)
    }
}
